subprojects/core/src/main/java/org/gradle/api/internal/attributes/JoinedAttributeContainer.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.attributes;
+
+import com.google.common.collect.Sets;
+import org.gradle.api.attributes.Attribute;
+import org.gradle.api.attributes.AttributeContainer;
+import org.gradle.api.provider.Provider;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Joins a parent attribute container to a child. Any attribute in the child container will
+ * override attributes in the parent container. All mutation operations are forwarded to
+ * the child container.
+ */
+public class JoinedAttributeContainer extends AbstractAttributeContainer {
```

Comment:
I agree `HeirarchicalAttributeContainer` is a better name, I was also thinking of that as an option -- Updated to that name. But, I think parent and child are quite fitting names for this relationship. Those terms are quite often used in hierarchical contexts. 

As for the method name in `ImmutableAttributesFactory`, I can't think of a parallel word for `join`. Maybe `override`? `hierarchy`? `hierarchify` (joke, but if this word existed it would be the right one IMO)?